### PR TITLE
feat(router): add `gateway_status_map` CRUD APIs

### PR DESCRIPTION
**Pull request by sai-harsha-vardhan**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
add `gateway_status_map` CRUD APIs

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
**Tested Manually**
1. GSM Create
CURL
`curl --location --request POST 'https://sandbox.hyperswitch.io/gsm' \
--header 'api-key: {ADMIN_API_KEY}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "connector": "stripe",
    "flow": "Authorize",
    "sub_flow": "sub_flow",
    "code": "incorrect_cvc",
    "message": "incorrect_cvc",
    "status": "failure",
    "decision": "retry",
    "step_up_possible": true
}'`
image
image

2. GSM Update
CURL
`curl --location --request POST 'https://sandbox.hyperswitch.io/gsm/update' \
--header 'api-key: {ADMIN_API_KEY}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "connector": "stripe",
    "flow": "Authorize",
    "sub_flow": "sub_flow",
    "code": "incorrect_cvc",
    "message": "incorrect_cvc",
    "decision": "do_default"
}'`
image

3. GSM Retrieve
CURL
`curl --location --request POST 'https://sandbox.hyperswitch.io/gsm/get' \
--header 'api-key: {ADMIN_API_KEY}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "connector": "stripe",
    "flow": "Authorize",
    "sub_flow": "sub_flow",
    "code": "incorrect_cvc",
    "message": "incorrect_cvc"
}'`
image

4. GSM Delete
CURL
`curl --location --request POST 'https://sandbox.hyperswitch.io/gsm/delete' \
--header 'api-key: {ADMIN_API_KEY}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "connector": "stripe",
    "flow": "Authorize",
    "sub_flow": "sub_flow",
    "code": "incorrect_cvc",
    "message": "incorrect_cvc"
}'`
image

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/routes/gsm.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        api_locking::LockAction::NotApplicable,
+    ))
+    .await
+}
+
+#[instrument(skip_all, fields(flow = ?Flow::GsmRuleRetrieve))]
+pub async fn get_gsm_rule(
+    state: web::Data<AppState>,
+    req: HttpRequest,
+    json_payload: web::Json<gsm_api_types::GsmRetrieveRequest>,
```
Comment (Narayanbhat166) [inline crates/router/src/routes/gsm.rs:36]: why is this a payload? do we not have a gsm identifier through which the gsm rule can be fetched?

File: crates/router/src/routes/app.rs [hunk 2]
```diff
@@ -666,6 +666,20 @@ impl BusinessProfile {
...
+
+#[cfg(feature = "olap")]
+impl Gsm {
+    pub fn server(state: AppState) -> Scope {
+        web::scope("/gsm")
+            .app_data(web::Data::new(state))
+            .service(web::resource("").route(web::post().to(create_gsm_rule)))
+            .service(web::resource("/get").route(web::post().to(get_gsm_rule)))
+            .service(web::resource("/update").route(web::post().to(update_gsm_rule)))
+            .service(web::resource("/delete").route(web::post().to(delete_gsm_rule)))
```
Comment (Narayanbhat166) [inline crates/router/src/routes/app.rs:679]: the endpoints and http methods do not match, ideally get should go with the GET http method, same for the rest

File: crates/router/src/routes/gsm.rs (hunk 1 above)
Comment (sai-harsha-vardhan) [reply crates/router/src/routes/gsm.rs:36]: We need to query GSM with connector, flow, sub_flow, code and message

File: crates/router/src/routes/app.rs (hunk 2 above)
Comment (sai-harsha-vardhan) [reply crates/router/src/routes/app.rs:679]: For all CRUD APIs we need payload with connector, flow, sub_flow, code and message fields

**Comment by Narayanbhat166** [comment]
In order to make it easier to use the gsm apis, a gsm list endpoint will be a good addition. This endpoint can have filters for connector, flow_type etc. Also the update endpoint should not allow the primary key fields to be updated and instead allow only the actions ( step_up, retry_action etc ) to be updated. Each gsm can have a unique identifier `gsm_abcd` which will act as a primary key to update / delete the gsm entry. 

This can be added as a good first issue.

**Review by Narayanbhat166 — DISMISSED** [review]

File: crates/router/src/routes/app.rs (hunk 2 above)
Comment (SanchithHegde) [reply crates/router/src/routes/app.rs:679]: The GET endpoint can use query parameters if the values are all single values (and not a list of values).

**Review by SanchithHegde — APPROVED** [review]
Other than that, looks good to me!

**Comment by likhinbopanna** [comment]
@sai-harsha-vardhan Can you add the cURL request?

**Comment by Narayanbhat166** [comment]
are the errors due to changes in this PR?

**Comment by sai-harsha-vardhan** [comment]
> are the errors due to changes in this PR?

@Narayanbhat166 This is due to APIEventMetrics which are added recently, fixing it!

**Review by SanchithHegde — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Pull request merged**